src/neo/Ledger/Blockchain.cs

```diff
@@ -498,13 +499,17 @@ private void Persist(Block block)
 
                     snapshot.Transactions.Add(tx.Hash, state);
 
-                    using (ApplicationEngine engine = new ApplicationEngine(TriggerType.Application, tx, snapshot.Clone(), tx.SystemFee))
+                    using (ApplicationEngine engine = new ApplicationEngine(TriggerType.Application, tx, clonedSnapshot, tx.SystemFee))
                     {
                         engine.LoadScript(tx.Script);
                         state.VMState = engine.Execute();
                         if (state.VMState == VMState.HALT)
                         {
-                            engine.Snapshot.Commit();
+                            clonedSnapshot.Commit();
```

Comment:
Commit at the end or when one FAULT?

Reply:
Yes, if everything's all right commit the changes just as original logic, or create a new clone for fault occasions.

Reply:
tx A successed ---> clonedSnapshot (don't commit, data not be saved)
tx B failed  ----> create a new cloneSnapshot, but this will discard A's data.